Implement wxSlider ticks under wxGTK

iwbnwif opened the pull request:
Implement wxSlider ticks under wxGTK

Tick marks were not available for wxSlider under GTK+ 2 or GTK+ 3
implementations of wxWidgets. However, tick marks have been available
for the GtkScale widget since GTK+ 2 version 2.16.

This change adds similar functionality in relation to tick marks
as was already available in wxMSW builds.

This has been tested using the 'samples/widgets' example under both GTK+ 2 and GTK+ 3 on Ubuntu 19.04.

There are a couple of limitations:
1. Ticks can only appear below (horizontal) and to the right (vertical) of the slider under GTK+ 2.
2. GTK+ 2 version 2.16 or later is required. It might be best to add a version check to avoid build errors with earlier versions of the toolkit?

vadz commented:
A couple of quick notes:

- To make a proper PR, please create a new branch from `origin/master` and then cherry-pick just the second commit to it and update this PR from this branch (i.e. force-push the new branch). To keep things simple, make sure your new branch has the same `slider-update` name. You will need to rename your existing branch using `git branch -m` to something else first.
- We definitely need GTK 2 version checks. This is tricky and I'd understand if you just implemented this for GTK 3 and leave 2 alone, but it's not that complex: you just need to have _both_ compile- and run-time checks. See `wxListBox::GetTopItem()` for a simple example.

TIA!

iwbnwif commented:
Thank you for the notes and git tips. I have created a new branch, synced it to origin/master and reapplied my edits.

Also, I have tried to follow the version checks (compile and run time).

But aarggh!! git has got me again. I tried to follow the forced push procedure, but got blocked because my branch is now behind. So I pull'd, not expecting the merge, so the changed files have ended up in this PR. I tried to revert the merge, but it's still here and the revert is added too.

If anyone can tell me how to get out of this tangle, I would be happy to tidy it up.

iwbnwif committed:
Implement wxSlider ticks under wxGTK

Tick marks were not available for wxSlider under GTK+ 2 or GTK+ 3
implementations of wxWidgets. However, tick marks have been available
for the GtkScale widget since GTK+ 2 version 2.16.

This change adds similar functionality in relation to tick marks
as was already available in wxMSW builds.

vadz commented on include/wx/gtk/slider.h:
```diff
@@ -9,6 +9,8 @@
 #ifndef _WX_GTK_SLIDER_H_
 #define _WX_GTK_SLIDER_H_
 
+#include "wx/gtk/private/wrapgtk.h"
```
As Travis CI build failures show, you can't include this private header from a public one. This means that you can't use `GTK_CHECK_VERSION()` here neither, you need to always declare the virtual functions but implement them differently depending on the version check.

iwbnwif replied:
Okay, this was a little awkward because it also means that I can't use native GTK+ structures (GtkPositionType in this case) as a member. I suppose it makes sense to decouple the toolkit specific implementation. 

As a result, I have added code to `wxSlider::SetTick()` to convert the style to a tick position 'on the fly'. This might not be quite optimal, but it is relatively simple and understandable. I am hoping that it is safe to use `GetWindowStyle()` here.

There are a couple of ways I can see of using `GTK_CHECK_VERSION()`:
1. Wrap the internal implementation of each function so that the functions compile as empty for older toolkit versions
2. Implement two versions of each function (one of which is empty), and use `GTK_CHECK_VERSION()` to determine which version is compiled.
The second option would result in less conditional directives as all functions could be wrapped in a single branch, but I have chosen the first simply because it is consistent with the way it has been done in `wxListBox::GetTopItem()`

vadz replied:
I think the current version is just fine, i.e. I agree both with converting the style to GTK enum on the fly and with using version checks inside the functions -- although, to be honest, I'd be fine with the other implementation too.

The only thing I might change would be to add `wxUnusedVar(param)` to avoid warnings about unused parameters with GTK < 2.16, but this is hardly critical, of course.

So if you don't have any objections, I'm going to merge this PR soon. Thanks again!

iwbnwif replied:
> The only thing I might change would be to add wxUnusedVar(param) to avoid warnings about unused parameters with GTK < 2.16, but this is hardly critical, of course.

Good spot, it's always best to have a clean compile. I've added them now, but cannot test as I don't have a GTK < 2.16 build environment.

I also tidied up the spaces around the 'if' statements.

I don't have any objections, please go ahead and merge if you think it's ready. Thanks!

vadz commented on src/gtk/slider.cpp:
```diff
@@ -297,6 +296,7 @@ bool wxSlider::Create(wxWindow *parent,
     m_scrollEventType = GTK_SCROLL_NONE;
     m_needThumbRelease = false;
     m_blockScrollEvent = false;
+    m_tickFreq = 0;
```
It looks like this is left uninitialized when default ctor is used, which is relatively harmless but still not ideal as `GetTickFreq()` will return garbage until `Create()` is called.

iwbnwif replied:
Sorry for the slow reply. Hmm, the only way I can see around this is to add the initialisation to both the default ctor and `Create()` because adding a default value inside the class declaration won't work on older compilers.

vadz replied:
The standard way used in wx is to have a private `Init()` method called from all ctors, including the default one.

iwbnwif replied:
Yay, thanks - good strategy!

vadz reviewed (changes requested):
The code and documentation changes look great, but the build must be fixed before this can be merged, could you please do it? TIA!

iwbnwif committed:
Implement wxSlider ticks under wxGTK

Tick marks were not available for wxSlider under GTK+ 2 or GTK+ 3
implementations of wxWidgets. However, tick marks have been available
for the GtkScale widget since GTK+ 2 version 2.16.

This change adds similar functionality in relation to tick marks
as was already available in wxMSW builds.

- Further commit to correct build errors and implement private
  Init() function to initialise member variables from any ctor.

iwbnwif committed:
Tick marks were not available for wxSlider under GTK+ 2 or GTK+ 3
implementations of wxWidgets. However, tick marks have been available
for the GtkScale widget since GTK+ 2 version 2.16.

This change adds similar functionality in relation to tick marks
as was already available in wxMSW builds.

- Further commit to add wxUnusedVar(var) to prevent compiler warnings
  and also add spaces within 'if' statements.

vadz commented:
Thanks again for your excellent work on this!

iwbnwif commented:
Thanks, now onwards with the spell checker :)
